pkg/services/signingkeys/signingkeystore/store.go

```diff
@@ -0,0 +1,194 @@
+	AddPrivateKey(ctx context.Context, keyID string, alg jose.SignatureAlgorithm, privateKey crypto.Signer, expiresAt *time.Time, force bool) error
+	// GetPrivateKey returns the private key with the specified key ID
+	GetPrivateKey(ctx context.Context, keyID string) (crypto.Signer, error)
+}
+
+var _ SigningStore = (*Store)(nil)
+
+type Store struct {
+	dbStore        db.DB
+	secretsService secrets.Service
+}
+
+type SigningKey struct {
+	ID         int64                   `json:"-" db:"id"`
+	KeyID      string                  `json:"key_id" db:"key_id"`
+	PrivateKey []byte                  `json:"private_key" db:"private_key"`
+	AddedAt    time.Time               `json:"added_at" db:"added_at"`
+	ExpiresAt  *time.Time              `json:"expires_at" db:"expires_at"`
+	Alg        jose.SignatureAlgorithm `json:"alg" db:"alg"`
+}
+
+func NewSigningKeyStore(dbStore db.DB, secretsService secrets.Service) *Store {
+	return &Store{
+		dbStore:        dbStore,
+		secretsService: secretsService,
+	}
+}
+
+func (s *Store) GetJWKS(ctx context.Context) (jose.JSONWebKeySet, error) {
+	keySet := jose.JSONWebKeySet{}
+
+	keys := []*SigningKey{}
+	err := s.dbStore.GetSqlxSession().Select(ctx, &keys, "SELECT * FROM signing_key")
+	if err != nil {
+		return keySet, err
+	}
+
+	for _, key := range keys {
+		assertedKey, err := s.decodePrivateKey(ctx, key)
+		if err != nil {
+			return keySet, err
+		}
+
+		keySet.Keys = append(keySet.Keys, jose.JSONWebKey{
+			Key:       assertedKey.Public(),
+			Algorithm: string(key.Alg),
+			KeyID:     key.KeyID,
+			Use:       "sig",
+		})
+	}
+
+	return keySet, nil
+}
+
+// AddPrivateKey adds a private key to the service.
+func (s *Store) AddPrivateKey(ctx context.Context,
+	keyID string, alg jose.SignatureAlgorithm, privateKey crypto.Signer, expiresAt *time.Time, force bool) error {
+	privateKeyPEM, err := s.encodePrivateKey(ctx, privateKey)
+	key := &SigningKey{
+		KeyID:      keyID,
+		PrivateKey: privateKeyPEM,
+		AddedAt:    time.Now(),
+		Alg:        alg,
+		ExpiresAt:  expiresAt,
+	}
+
+	dbSession := s.dbStore.GetSqlxSession()
+	err = dbSession.WithTransaction(ctx, func(tx *session.SessionTx) error {
+		existingKey := SigningKey{}
+		err := tx.Get(ctx, &existingKey, "SELECT * FROM signing_key WHERE key_id = ?", keyID)
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return err
+		}
+
+		if len(existingKey.PrivateKey) == 0 {
+			_, err = tx.Exec(ctx,
+				"INSERT INTO signing_key (key_id, private_key, added_at, alg, expires_at) VALUES (?, ?, ?, ?, ?)",
+				key.KeyID, key.PrivateKey, key.AddedAt, key.Alg, key.ExpiresAt)
+			return err
+		}
+
+		if force || (existingKey.ExpiresAt != nil && existingKey.ExpiresAt.Before(time.Now())) {
+			_, err = tx.Exec(ctx,
+				"UPDATE signing_key SET private_key = ?, added_at = ?, alg = ?, expires_at = ? WHERE key_id = ?",
+				key.PrivateKey, key.AddedAt, key.Alg, key.ExpiresAt, key.KeyID)
+			return err
+		}
```

Comment:
I guess this is intended to rotate keys right?

When a key rotation happen we should probably keep the old key for a while so we don't immateriality invalidate everything signed with the previous key, WDYT?

Reply:
+1, maybe introducing a `version` field to the `signing_key` table? (And eventually the JWT header's `kid` should contain the version as well, something like `keyname-1`)

Reply:
+1, that's a good point, we need to account for both keys existing at the same time. But I guess we'll never rotate them (no update), as they need to exist as independent entries. Let me see if I can find a way to make it work